### PR TITLE
feat(p2d): enhance testcase reading

### LINKED ISSUE #5: P2D: Enhance testcase reading
For my two problem.xml, there are:

```xml
    <judging cpu-name="Intel(R) Core(TM) i3-8100 CPU @ 3.60GHz" cpu-speed="3600" input-file="" output-file="">
        <testset name="tests">
            <time-limit>1000</time-limit>
            <memory-limit>268435456</memory-limit>
            <test-count>11</test-count>
            <input-path-pattern>tests/%02d</input-path-pattern>
            <answer-path-pattern>tests/%02d.a</answer-path-pattern>
            <tests>
                <test method="manual" sample="true"/>
                <test cmd="gen 0 10" method="generated"/>
                <test cmd="gen 495 505" method="generated"/>
                <test cmd="gen 990 1000" method="generated"/>
                <test cmd="gen 0 100" method="generated"/>
                <test cmd="gen 450 550" method="generated"/>
                <test cmd="gen 900 1000" method="generated"/>
                <test cmd="gen 0 1000 a" method="generated"/>
                <test cmd="gen 0 1000 b" method="generated"/>
                <test cmd="gen 0 1000 c" method="generated"/>
                <test cmd="gen 0 1000 d" method="generated"/>
            </tests>
        </testset>
    </judging>
```

and

```xml
    <judging cpu-name="Intel(R) Core(TM) i3-8100 CPU @ 3.60GHz" cpu-speed="3600" input-file="" output-file="">
        <testset name="tests">
            <time-limit>1000</time-limit>
            <memory-limit>268435456</memory-limit>
            <test-count>20</test-count>
            <input-path-pattern>tests/%02d</input-path-pattern>
            <answer-path-pattern>tests/%02d.a</answer-path-pattern>
            <tests>
                <test method="manual" sample="true"/>
                <test method="manual" sample="true"/>
                <test method="manual" sample="true"/>
                <test method="manual"/>
                <test method="manual"/>
                <test method="manual"/>
                <test method="manual"/>
                <test method="manual"/>
                <test method="manual"/>
                <test method="manual"/>
                <test method="manual"/>
                <test method="manual"/>
                <test method="manual"/>
                <test method="manual"/>
                <test method="manual"/>
                <test method="manual"/>
                <test method="manual"/>
                <test method="manual"/>
                <test method="manual"/>
                <test method="manual"/>  
            </tests>
        </testset>
    </judging>
```

It can be seen that the sample property is `sample="true"` in the tag attribute.

And we can know the testcase pattern is `tests/%02d` and `tests/%02d.a`.

For those generated testcases, you can put the `cmd` attribute value with prefix `[GEN] ` into `data/secret/01.desc` to be a testcase description.

### PR DESCRIPTION
**Pull request by cubercsl**
fix #5

### CODE REVIEW

File: p2d/p2d.py [hunk 1]
```diff
@@ -61,32 +79,21 @@ def write_ini(probid, timelimit, color):
...
-                return None
-            checker_source = checker.find('source')
-            checker_file = '{}/{}'.format(package_dir, checker_source.attrib['path'])
-            with open(checker_file, 'r', encoding='utf-8') as f:
-                file_md5 = hashlib.md5(f.read().replace('\r\n', '\n').encode('utf-8'))
-            return file_md5.hexdigest().lower()
-
-        checker_md5 = get_checker_md5(checker)
-        logger.info('Checker md5: {}'.format(checker_md5))
+        checker_name = checker.attrib['name']
```
Comment (yang-er) [inline p2d/p2d.py]: image

There's no `name` attribute for self-defined checker.

```xml
        <checker type="testlib">
            <source path="files/chk.cpp" type="cpp.g++17"/>
            <copy path="check.cpp" type="cpp.g++17"/>
            <testset>
                <test-count>4</test-count>
                <input-path-pattern>files/tests/checker-tests/%02d</input-path-pattern>
                <output-path-pattern>files/tests/checker-tests/%02d.o</output-path-pattern>
                <answer-path-pattern>files/tests/checker-tests/%02d.a</answer-path-pattern>
                <tests>
                    <test verdict="ok"/>
                    <test verdict="crashed"/>
                    <test verdict="wrong-answer"/>
                    <test verdict="ok"/>
                </tests>
            </testset>
        </checker>
```

File: p2d/p2d.py [hunk 2]
```diff
@@ -39,16 +47,26 @@ def parse_problem():
...
+        timelimit = int(testset.find('time-limit').text) / 1000.0
+        memlimit = int(testset.find('memory-limit').text) // 1048576
         checker = root.find('assets').find('checker')
         interactor = root.find('assets').find('interactor')
+        input_path_pattern =testset.find('input-path-pattern').text
+        answer_path_pattern = testset.find('answer-path-pattern').text 
+        tests = []
+        for test in testset.find('tests').findall('test'):
+            method = test.attrib['method']
+            cmd = test.attrib.get('cmd', None)
```
Comment (yang-er) [inline p2d/p2d.py:117]:
```xml
<test description="File &quot;test.in&quot;" method="manual"/>
```

For manual types, you can read the `description` attribute and put it into the `.desc`.

**Comment by yang-er** [comment]
Nice PR!

**Comment by cubercsl** [comment]
Part of the code is refactored to facilitate other scripts to call.

**Pull request merged**